src/Microsoft.DotNet.Interactive.ExtensionLab/MsSqlKernel.cs

```diff
@@ -3,21 +3,219 @@
 
 using System.Data.Common;
 using Microsoft.Data.SqlClient;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.Events;
+using System.Data;
+using System.Reactive.Disposables;
+using System.Threading;
+using Microsoft.DotNet.Interactive.Formatting;
+using Newtonsoft.Json.Linq;
+using System.Collections.Concurrent;
+using System.Text;
+using Microsoft.DotNet.Interactive.Utility;
 
 namespace Microsoft.DotNet.Interactive.ExtensionLab
 {
-    public class MsSqlKernel : SqlKernel
+    public class MsSqlKernel :
+        Kernel,
+        IKernelCommandHandler<SubmitCode>,
+        IKernelCommandHandler<RequestCompletions>,
+        IKernelCommandHandler<RequestHoverText>
     {
+        private bool _connected = false;
+        private bool _intellisenseReady = false;
+        private readonly string _tempFilePath;
+        private readonly string _tempFileUri;
         private readonly string _connectionString;
+        private readonly MsSqlServiceClient _serviceClient;
+
+        private TaskCompletionSource<ConnectionCompleteParams> _connectionCompleted = new TaskCompletionSource<ConnectionCompleteParams>();
+
+        private Func<QueryCompleteParams, Task> _queryCompletionHandler = null;
 
         public MsSqlKernel(string name, string connectionString) : base(name)
         {
+            _tempFilePath = Path.GetTempFileName();
+            _tempFileUri = MsSqlServiceClient.GetUriForFilePath(_tempFilePath);
             _connectionString = connectionString;
+
+            _serviceClient = MsSqlServiceClient.Instance;
+            _serviceClient.Initialize();
+
+            _serviceClient.OnConnectionComplete += HandleConnectionComplete;
+            _serviceClient.OnQueryComplete += HandleQueryCompleteAsync;
+            _serviceClient.OnIntellisenseReady += HandleIntellisenseReady;
+
+            RegisterForDisposal(() =>
+            {
+                if (_connected)
+                {
+                    _serviceClient.DisconnectAsync(_tempFileUri).Wait();
+                }
+            });
+            RegisterForDisposal(() => File.Delete(_tempFilePath));
+        }
+
+        private void HandleConnectionComplete(object sender, ConnectionCompleteParams connParams)
+        {
+            if (connParams.OwnerUri.Equals(_tempFileUri))
+            {
+                if (connParams.ErrorMessage != null)
+                {
+                    _connectionCompleted.SetException(new Exception(connParams.ErrorMessage));
```

Comment:
It's generally a good idea to avoid using `Exception`. Is there a more specific exception we could use here? What are some examples of why an error will occur?

Reply:
This ErrorMessage can come from any type of Exception. We just catch the base Exception class on the service side for reporting the error message: https://github.com/microsoft/sqltoolsservice/blob/main/src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs#L565

Reply:
Would `SqlException` be appropriate? A custom exception type?